README.md
# Flutter Launcher Icons Maker

Fork from Flutter Launcher Icons. Add merge pull request of generate windows, macos and web icons.

A command-line tool which simplifies the task of updating your Flutter app's launcher icon. Fully flexible, allowing you to choose what platform you wish to update the launcher icon for and if you want, the option to keep your old launcher icon in case you want to revert back sometime in the future.


## :sparkles: What's New

[Changelog](https://github.com/gsmlg-dev/flutter_launcher_icons_maker/blob/master/CHANGELOG.md).

## :book: Guide

### 1. Setup the config file

Run the following command to create a new config automatically:

```shell
flutter pub run flutter_launcher_icons:generate
```

This will create a new file called `flutter_launcher_icons.yaml` in your `flutter` project's root directory.

If you want to override the default location or name of the config file, use the `-f` flag:

```shell
flutter pub run flutter_launcher_icons:generate -f <your config file name here>
```

To override an existing config file, use the `-o` flag:


```shell
flutter pub run flutter_launcher_icons:generate -o
```

OR

Add your Flutter Launcher Icons configuration to your `pubspec.yaml`.  
An example is shown below. More complex examples [can be found in the example projects](https://github.com/fluttercommunity/flutter_launcher_icons/tree/master/example).

```yaml
dev_dependencies:
  flutter_launcher_icons: "^0.13.1"
flutter_launcher_icons:
  android: "launcher_icon"
  ios: true
  macos: true
  windows: true
  web: true
  image_path: "assets/icon/icon.png"
  min_sdk_android: 21 # android min sdk min:16, default 21
  web:
    generate: true
    image_path: "path/to/image.png"
    background_color: "#hexcode"
    theme_color: "#hexcode"
  windows:
    generate: true
    image_path: "path/to/image.png"
    icon_size: 48 # min:48, max:256, default: 48
  macos:
    generate: true
    image_path: "path/to/image.png"
```

```shell
flutter pub get
flutter pub run flutter_launcher_icons_maker:main -f <your config file name here>
```

Note: If you are not using the existing `pubspec.yaml` ensure that your config file is located in the same directory as it.


### 2. Run the package


After setting up the configuration, all that is left to do is run the package.

```shell
flutter pub get
flutter pub run flutter_launcher_icons
```

If you name your configuration file something other than `flutter_launcher_icons.yaml` or `pubspec.yaml` you will need to specify
the name of the file when running the package.

```shell
flutter pub get
flutter pub run flutter_launcher_icons -f <your config file name here>
```

Note: If you are not using the existing `pubspec.yaml` ensure that your config file is located in the same directory as it.

If you encounter any issues [please report them here](https://github.com/fluttercommunity/flutter_launcher_icons/issues).

In the above configuration, the package is setup to replace the existing launcher icons in both the Android and iOS project
with the icon located in the image path specified above and given the name "launcher_icon" in the Android project and "Example-Icon" in the iOS project.

## :mag: Attributes

Shown below is the full list of attributes which you can specify within your Flutter Launcher Icons configuration.

### Global

- `image_path`: The location of the icon image file which you want to use as the app launcher icon.

### Android

- `android`
  - `true`: Override the default existing Flutter launcher icon for the platform specified
  - `false`: Ignore making launcher icons for this platform
  - `icon/path/here.png`: This will generate a new launcher icons for the platform with the name you specify, without removing the old default existing Flutter launcher icon.
- `image_path`: The location of the icon image file which you want to use as the app launcher icon
- `image_path_android`: The location of the icon image file specific for Android platform (optional - if not defined then the image_path is used)
- `min_sdk_android`: Specify android min sdk value
**The next two attributes are only used when generating Android launcher icon**

- `adaptive_icon_background`: The color (E.g. `"#ffffff"`) or image asset (E.g. `"assets/images/christmas-background.png"`) which will
be used to fill out the background of the adaptive icon.
- `adaptive_icon_foreground`: The image asset which will be used for the icon foreground of the adaptive icon
*Note: Adaptive Icons will only be generated when both adaptive_icon_background and adaptive_icon_foreground are specified. (the image_path is not automatically taken as foreground)*

### IOS

- `ios`
  - `true`: Override the default existing Flutter launcher icon for the platform specified
  - `false`: Ignore making launcher icons for this platform
  - `icon/path/here.png`: This will generate a new launcher icons for the platform with the name you specify, without removing the old default existing Flutter launcher icon.
- `image_path_ios`: The location of the icon image file specific for iOS platform (optional - if not defined then the image_path is used)
- `remove_alpha_ios`: Removes alpha channel for IOS icons
- `background_color_ios`: The color (in the format "#RRGGBB") to be used as the background when removing the alpha channel. It is used only when the `remove_alpha_ios` property is set to true. (optional - if not defined then `#ffffff` is used)

### Web

- `web`: Add web related configs
  - `generate`: Specifies weather to generate icons for this platform or not
  - `image_path`: Path to web icon.png
  - `background_color`: Updates *background_color* in `web/manifest.json`
  - `theme_color`: Updates *theme_color* in `web/manifest.json`

### Windows

- `windows`: Add Windows related configs
  - `generate`: Specifies weather to generate icons for Windows platform or not
  - `image_path`: Path to web icon.png
  - `icon_size`: Windows app icon size. Icon size should be within this constrains *48<=icon_size<=256, defaults to 48*
  
### MacOS

- `macos`: Add MacOS related configs
  - `generate`: Specifies weather to generate icons for MacOS platform or not
  - `image_path`: Path to macos icon.png file

*Note: iOS icons should [fill the entire image](https://stackoverflow.com/questions/26014461/black-border-on-my-ios-icon) and not contain transparent borders.*

## Flavor support

Create a Flutter Launcher Icons configuration file for your flavor. The config file is called `flutter_launcher_icons-<flavor>.yaml` by replacing `<flavor>` by the name of your desired flavor.

The configuration file format is the same.

An example project with flavor support enabled [has been added to the examples](https://github.com/fluttercommunity/flutter_launcher_icons/tree/master/example/flavors).

## :question: Troubleshooting

Listed a couple common issues with solutions for them

### Generated icon color is different from the original icon

Caused by an update to the image dependency which is used by Flutter Launcher Icons.

```txt
Use #AARRGGBB for colors instead of #AABBGGRR, to be compatible with Flutter image class.
```

[Related issue](https://github.com/fluttercommunity/flutter_launcher_icons/issues/98)

### Image foreground is too big / too small

For best results try and use a foreground image which has padding much like [the one in the example](https://github.com/fluttercommunity/flutter_launcher_icons/blob/master/example/default_example/assets/images/icon-foreground-432x432.png).

[Related issue](https://github.com/fluttercommunity/flutter_launcher_icons/issues/96)

### Dependency incompatible

You may receive a message similar to the following

```log
Because flutter_launcher_icons >=0.9.0 depends on args 2.0.0 and flutter_native_splash 1.2.0 depends on args ^2.1.1, flutter_launcher_icons >=0.9.0 is incompatible with flutter_native_splash 1.2.0.
And because no versions of flutter_native_splash match >1.2.0 <2.0.0, flutter_launcher_icons >=0.9.0 is incompatible with flutter_native_splash ^1.2.0.
So, because enstack depends on both flutter_native_splash ^1.2.0 and flutter_launcher_icons ^0.9.0, version solving failed.
pub get failed (1; So, because enstack depends on both flutter_native_splash ^1.2.0 and flutter_launcher_icons ^0.9.0, version solving failed.)
```

For a quick fix, you can temporarily override all references to a dependency: [See here for an example](https://github.com/fluttercommunity/flutter_launcher_icons/issues/262#issuecomment-879872076).

## :eyes: Example

[![Video Example](https://i.imgur.com/R28hqdz.png)](https://www.youtube.com/watch?v=RjNAxwcP3Tc)

Note: This is showing a very old version (v0.0.5)

### Special thanks

- Thanks to Brendan Duncan for the underlying [image package](https://pub.dev/packages/image) to transform the icons.
- Big thank you to all the contributors to the project. Every PR / reported issue is greatly appreciated!


CHANGELOG.md
# Changelog

## 0.13.1 (15th April 2023)

- Can now use `flutter_launcher_icons` instead of `flutter_icons` [#478](https://github.com/fluttercommunity/flutter_launcher_icons/pull/478)
- Can use command `flutter pub run flutter_launcher_icons:generate` to automatically generate config file [#475](https://github.com/fluttercommunity/flutter_launcher_icons/pull/475)


## 0.13.0 (7th April 2023)

- Fix remove alpha for iOS [#464](https://github.com/fluttercommunity/flutter_launcher_icons/pull/464)
- Updating code style [#472](https://github.com/fluttercommunity/flutter_launcher_icons/pull/472)
- Updated out of bounds dependency [#473](https://github.com/fluttercommunity/flutter_launcher_icons/pull/473)

## 0.12.0 (24th February 2023)

- Updated image package and other packages [#447](https://github.com/fluttercommunity/flutter_launcher_icons/pull/447)

## 0.11.0 (27th September 2022)
    
- Support for Macos Icons [#407](https://github.com/fluttercommunity/flutter_launcher_icons/pull/407)
- Cli-improvement [#400](https://github.com/fluttercommunity/flutter_launcher_icons/pull/400)
- Add `repository` and `issue_tracker` [#411](https://github.com/fluttercommunity/flutter_launcher_icons/pull/411) (thanks to [@patelpathik](https://github.com/patelpathik))
- Fix indent in web/manifest.json [#407](https://github.com/fluttercommunity/flutter_launcher_icons/pull/407)
- Fix the icons 50 and 57 in `contents.json` [#412](https://github.com/fluttercommunity/flutter_launcher_icons/pull/412) (thanks to [@adnanjpg](https://github.com/adnanjpg))
- Fix typos [#405](https://github.com/fluttercommunity/flutter_launcher_icons/pull/405) (thanks to [@edwardmp](https://github.com/edwardmp))
- Added newline to EOF [#325](https://github.com/fluttercommunity/flutter_launcher_icons/pull/325) (thanks to [@sandersaelmans](https://github.com/sandersaelmans))

## 0.10.0 (2nd August 2022)

- Support for Web Icons [#374](https://github.com/fluttercommunity/flutter_launcher_icons/pull/374)
- Support for Windows Icons [#382](https://github.com/fluttercommunity/flutter_launcher_icons/pull/382)
- Added missing IOS icon sizes [#298](https://github.com/fluttercommunity/flutter_launcher_icons/pull/298)
- Added `min_sdk_android` option [#392](https://github.com/fluttercommunity/flutter_launcher_icons/pull/392)
- Added documentation for `remove_alpha_ios` [#392](https://github.com/fluttercommunity/flutter_launcher_icons/pull/392)
- Fixed issue with loading config from `pubspec.yaml` [#398](https://github.com/fluttercommunity/flutter_launcher_icons/pull/398) (thanks to [@p-mazhnik](https://github.com/p-mazhnik))

## 0.9.3 (6th June 2022)

- Fixes to make sure it works for Flutter v2.8 (thanks to @RatakondalaArun)
- Fixed issue with incorrect version being shown

## 0.9.2 (22nd August 2021)

- Fixed issue where success message printed even when exception occured (thanks to @happy-san)

## 0.9.1 (25th July 2021)

- Upgrade args dependency to ^2.1.1 (thanks to @PiN73 and @comlaterra)
- Upgraded `image` and `test` dependencies

## 0.9.0 (28th Feb 2021)

- Null-safety support added (thanks to @SteveAlexander)
- Added option to remove alpha channel for iOS icons (thanks to @SimonIT)

## 0.8.1 (2nd Oct 2020)

- Fixed flavor support on windows (@slightfoot)

## 0.8.0 (12th Sept 2020)

- Added flavours support (thanks to @sestegra & @jorgecoca)
- Removed unassigned iOS icons (thanks to @melvinsalas)
- Fixing formatting (thanks to @mreichelt)

## 0.7.5 (24th April 2020)

- Fixed issue where new lines were added to Android manifest (thanks to @mreichelt)
- Improvements to code quality and general tidying up (thanks to @connectety)
- Fixed Android example project not running (needed to be migrated to AndroidX)

## 0.7.4 (28th Oct 2019)

- Worked on suggestions from [pub.dev](https://pub.dev/packages/flutter_launcher_icons#-analysis-tab-)

## 0.7.3 (3rd Sept 2019)

- Lot of refactoring and improving code quality (thanks to @connectety)
- Added correct App Store icon settings (thanks to @richgoldmd)

## 0.7.2 (25th May 2019)

- Reverted back using old interpolation method

## 0.7.1 (24th May 2019)

- Fixed issue with image dependency not working on latest version of Flutter (thanks to @sboutet06)
- Fixed iOS icon sizes which were incorrect (thanks to @sestegra)
- Removed dart_config git dependency and replaced with yaml dependency
- Refactoring of code

## 0.7.0 (22nd November 2018)

- Now ensuring that the Android file name is valid - An error will be thrown if it doesn't meet the criteria
- Fixed issue where there was a git diff when there was no change
- Fixed issue where iOS icon would be generated when it shouldn't be
- Added support for drawables to be used for adaptive icon backgrounds
- Added support for Flutter Launcher Icons to be able to run with it's own config file (no longer necessary to add to pubspec.yaml)

## 0.6.1 (26th August 2018)

- Upgraded test package
- Due to issue with dart_config not working with Dart 2.1.0, now using forked version of dart_config which contains fixes from both @v3rm0n and @SPodjasek

## 0.6.0 (8th August 2018)

- Moved the package to [Flutter Community](https://github.com/fluttercommunity/community)

## 0.5.2 (19th June 2018)

- Previous release didn't fix adaptive icons, just prevented the error message from appearing. This should hopefully fix it!

## 0.5.1 (18th June 2018)

- Fix for adaptive icons

## 0.5.0 (12th June 2018)

- [Android] Support for adaptive icons added (Suggestion #23)

## 0.4.0 (9th June 2018)

- Now possible to generate icons for each platform with different image paths (one for iOS icon and a separate one for Android)

## 0.3.3 (28th May 2018)

- Upgraded dart image package dependency to 2.0.0 (issue #26)

## 0.3.2 (2nd May 2018)

- Bug fixing

## 0.3.1 (1st May 2018)

- Bug fixing

## 0.3.0 (1st May 2018)

- Fixed issue where icons produced weren't the correct size (Due to images not with a 1:1 aspect r    ation)

test/web_test.dart
import 'dart:io';

import 'package:flutter_launcher_icons_maker/icon_template.dart';
import 'package:flutter_launcher_icons_maker/web.dart';
import 'package:image/image.dart';
import 'package:test/test.dart';

const String wideImagePath = 'test/images/wide.png';
const String tallImagePath = 'test/images/tall.png';
const String squareImagePath = 'test/images/square.png';

void main() {
  test('Has favicon support.', () {
    bool hasFavicon = false;

    for (IconTemplate template in webIcons) {
      hasFavicon = template.name == 'favicon.png';
    }

    expect(hasFavicon, true);
  });

  test('Correctly crops wide input images', () {
    testImageCrop(wideImagePath, 192);
  });

  test('Correctly crops tall input images', () {
    testImageCrop(tallImagePath, 192);
  });

  test('Correctly crops square input images.', () {
    testImageCrop(squareImagePath, 192);
  });
}

void testImageCrop(String imagePath, int expectedSize) {
  final IconTemplate testTemplate =
      iconGenerator.get(name: 'Icon-$expectedSize.png', size: expectedSize);
  final Image? testImage = decodeImage(File(imagePath).readAsBytesSync());
  expect(testImage, isNotNull);

  final Image resizedImage = testTemplate.createFrom(testImage!);

  expect(resizedImage.width == resizedImage.height, true);
  expect(resizedImage.width, expectedSize);
}


test/main_test.dart
import 'dart:io';

import 'package:args/args.dart';
import 'package:flutter_launcher_icons_maker/android.dart' as android;
import 'package:flutter_launcher_icons_maker/constants.dart' as constants;
import 'package:flutter_launcher_icons_maker/ios.dart' as ios;
import 'package:flutter_launcher_icons_maker/main.dart' as main_dart;
import 'package:path/path.dart' as pathlib;
import 'package:test/test.dart';
import 'package:yaml/yaml.dart' as yaml;

// Unit tests for main.dart
void main() {
  test('iOS icon list is correct size', () {
    expect(ios.iosIcons.length, 21);
  });

  test('Android icon list is correct size', () {
    expect(android.androidIcons.length, 5);
  });

  test(
      'iOS image list used to generate Contents.json for icon directory is correct size',
      () {
    expect(ios.createImageList('blah').length, 25);
  });

  group('config file from args', () {
    // Create mini parser with only the wanted option, mocking the real one
    final ArgParser parser = ArgParser()
      ..addOption(constants.fileOption, abbr: 'f');
    final String testDir = pathlib.join(
        '.dart_tool', 'flutter_launcher_icons', 'test', 'config_file');
      ..addOption(
        main_dart.fileOption,
        abbr: 'f',
        defaultsTo: defaultConfigFile,
      )
      ..addOption(
        main_dart.prefixOption,
        abbr: 'p',
        defaultsTo: '.',
      );
    final String testDir =
        join('.dart_tool', 'flutter_launcher_icons', 'test', 'config_file');

    String currentDirectory = testDir;
    Future<void> setCurrentDirectory(String path) async {
      currentDirectory = pathlib.join(testDir, path);
      await Directory(currentDirectory).create(recursive: true);
    }

    File getFile(String filename) {
      final String path = pathlib.join(currentDirectory, filename);

      return File(path);
    }

    Map<String, dynamic>? loadConfig(ArgResults args) {
      return main_dart.loadConfigFileFromArgResults(args,
          cwd: currentDirectory);
    }

    setUp(() {
      currentDirectory = Directory.current.path;
    });
    tearDown(() {
      Directory.current = currentDirectory;
    });

    test('default', () async {
      await setCurrentDirectory('default');
      await File('flutter_launcher_icons.yaml').writeAsString('''
flutter_launcher_icons:
  android: true
  ios: false
''');
      final ArgResults argResults = parser.parse(<String>[]);
      final Config? config = main_dart.loadConfigFileFromArgResults(argResults);
      expect(config, isNotNull);
      expect(config!.android, isTrue);
    });

    test('default_use_pubspec', () async {
      await setCurrentDirectory('pubspec_only');
      await File('pubspec.yaml').writeAsString('''
flutter_launcher_icons:
  android: true
  ios: false
''');
      ArgResults argResults = parser.parse(<String>[]);
      final Config? config = main_dart.loadConfigFileFromArgResults(argResults);
      expect(config, isNotNull);
      expect(config!.ios, isFalse);

      // read pubspec if provided file is not found
      argResults = parser.parse(<String>['-f', defaultConfigFile]);
      expect(main_dart.loadConfigFileFromArgResults(argResults), isNotNull);
    });

    test('custom', () async {
      await setCurrentDirectory('custom');
      await File('custom.yaml').writeAsString('''
flutter_launcher_icons:
  android: true
  ios: true
''');
      // if no argument set, should fail
      ArgResults argResults = parser.parse(<String>['-f', 'custom.yaml']);
      final Config? config = main_dart.loadConfigFileFromArgResults(argResults);
      expect(config, isNotNull);
      expect(config!.ios, isTrue);

      // should fail if no argument
      argResults = parser.parse(<String>[]);
      expect(loadConfig(argResults), <String, dynamic>{});

      // or missing file
      argResults = parser.parse(<String>['-f', 'missing_custom.yaml']);
      expect(loadConfig(argResults), <String, dynamic>{});
    });
  });

  test('image_path is in config', () {
    final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'android': true,
      'ios': true,
      'web': true,
    };
    final config = Config.fromJson(flutterIconsConfig);
    expect(
      config.getImagePathAndroid(),
      equals('assets/images/icon-710x599.png'),
    );
    expect(config.getImagePathIOS(), equals('assets/images/icon-710x599.png'));
    final Map<String, dynamic> flutterIconsConfigAndroid = <String, dynamic>{
      'image_path_android': 'assets/images/icon-710x599.png',
      'android': true,
      'ios': true
    };
    final configAndroid = Config.fromJson(flutterIconsConfigAndroid);
    expect(
      configAndroid.getImagePathAndroid(),
      equals('assets/images/icon-710x599.png'),
    );
    expect(configAndroid.getImagePathIOS(), isNull);
    final Map<String, dynamic> flutterIconsConfigBoth = <String, dynamic>{
      'image_path_android': 'assets/images/icon-android.png',
      'image_path_ios': 'assets/images/icon-ios.png',
      'android': true,
      'ios': true
    };
    final configBoth = Config.fromJson(flutterIconsConfigBoth);
    expect(
      configBoth.getImagePathAndroid(),
      equals('assets/images/icon-android.png'),
    );
    expect(configBoth.getImagePathIOS(), equals('assets/images/icon-ios.png'));
  });

  test('At least one platform is in config file', () {
    final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'android': true,
      'ios': true,
      'web': true,
    };
    final config = Config.fromJson(flutterIconsConfig);
    expect(config.hasPlatformConfig, isTrue);
  });

  test('No platform specified in config', () {
    final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png'
    };
    final config = Config.fromJson(flutterIconsConfig);
    expect(config.hasPlatformConfig, isFalse);
  });

  test('No new Android icon needed - android: false', () {
    final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'android': false,
      'ios': true
    };
    final config = Config.fromJson(flutterIconsConfig);
    expect(config.isNeedingNewAndroidIcon, isFalse);
  });

  test('No new Android icon needed - no Android config', () {
    final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'ios': true
    };
    final config = Config.fromJson(flutterIconsConfig);
    expect(config.isNeedingNewAndroidIcon, isFalse);
  });

  test('No new iOS icon needed - ios: false', () {
    final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'android': true,
      'ios': false
    };
    final config = Config.fromJson(flutterIconsConfig);
    expect(config.isNeedingNewIOSIcon, isFalse);
  });

  test('No new web icon needed -- web: false', () {
    final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'web': false,
      'ios': true
    };
    final config = Config.fromJson(flutterIconsConfig);
    expect(config.isNeedingNewIOSIcon, isFalse);
  });
}


test/android_test.dart
import 'dart:io';

import 'package:flutter_launcher_icons_maker/android.dart' as android;
import 'package:flutter_launcher_icons_maker/config/config.dart';
import 'package:flutter_launcher_icons_maker/constants.dart';
import 'package:test/test.dart';

// unit tests for android.dart
void main() {
  test('Adaptive icon mipmap path is correct', () {
    const String path1 = 'android/app/src/main/res/';
    const String path2 = 'mipmap-anydpi-v26/';
    expect(android.isCorrectMipmapDirectoryForAdaptiveIcon(path1), false);
    expect(android.isCorrectMipmapDirectoryForAdaptiveIcon(path2), false);
    expect(
      android.isCorrectMipmapDirectoryForAdaptiveIcon(
        androidAdaptiveXmlFolder(null),
      ),
      true,
    );
  });

  test('Correct number of adaptive foreground icons', () {
    expect(android.adaptiveForegroundIcons.length, 5);
  });

  test('Correct number of android launcher icons', () {
    expect(android.androidIcons.length, 5);
  });

  test('Config contains string for generating new launcher icons', () {
    final Map<String, dynamic> flutterIconsConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'android': true,
      'ios': true
    };
    expect(
      Config.fromJson(flutterIconsConfig).isCustomAndroidFile,
      isFalse,
    );

    final Map<String, dynamic> flutterIconsNewIconConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'android': 'New Icon',
      'ios': true
    };
    expect(
      Config.fromJson(flutterIconsNewIconConfig).isCustomAndroidFile,
      isTrue,
    );
  });

  test('Prioritise image_path_android over image_path', () {
    final Map<String, dynamic> flutterIconsNewIconConfig = <String, dynamic>{
      'image_path': 'assets/images/icon-710x599.png',
      'image_path_android': 'assets/images/icon-android.png',
      'android': 'New Icon',
      'ios': true
    };
    expect(
      Config.fromJson(flutterIconsNewIconConfig).getImagePathAndroid(),
      equals('assets/images/icon-android.png'),
    );
  });

  test('Transforming manifest without icon must add icon', () async {
    final String inputManifest =
        getAndroidManifestExample('android:icon="@mipmap/ic_launcher"');
    final String expectedManifest =
        getAndroidManifestExample('android:icon="@mipmap/ic_other_icon_name"');

    await withTempFile('AndroidManifest.xml', (File androidManifestFile) async {
      androidManifestFile.writeAsStringSync(inputManifest);
      await android.overwriteAndroidManifestWithNewLauncherIcon(
        'ic_other_icon_name',
        androidManifestFile,
      );
      expect(androidManifestFile.readAsStringSync(), equals(expectedManifest));
    });
  });

  test(
      'Transforming manifest with icon already in place should leave it unchanged',
      () async {
    final String inputManifest =
        getAndroidManifestExample('android:icon="@mipmap/ic_launcher"');
    final String expectedManifest =
        getAndroidManifestExample('android:icon="@mipmap/ic_launcher"');

    await withTempFile('AndroidManifest.xml', (File androidManifestFile) async {
      androidManifestFile.writeAsStringSync(inputManifest);
      await android.overwriteAndroidManifestWithNewLauncherIcon(
        'ic_launcher',
        androidManifestFile,
      );
      expect(androidManifestFile.readAsStringSync(), equals(expectedManifest));
    });
  });

  test(
      'Transforming manifest with trailing newline should keep newline untouched',
      () async {
    final String inputManifest =
        getAndroidManifestExample('android:icon="@mipmap/ic_launcher"') + '\n';
    final String expectedManifest = inputManifest;

    await withTempFile('AndroidManifest.xml', (File androidManifestFile) async {
      androidManifestFile.writeAsStringSync(inputManifest);
      await android.overwriteAndroidManifestWithNewLauncherIcon(
        'ic_launcher',
        androidManifestFile,
      );
      expect(androidManifestFile.readAsStringSync(), equals(expectedManifest));
    });
  });

  test(
      'Transforming manifest with 3 trailing newlines should keep newlines untouched',
      () async {
    final String inputManifest =
        getAndroidManifestExample('android:icon="@mipmap/ic_launcher"') +
            '\n\n\n';
    final String expectedManifest = inputManifest;

    await withTempFile('AndroidManifest.xml', (File androidManifestFile) async {
      androidManifestFile.writeAsStringSync(inputManifest);
      await android.overwriteAndroidManifestWithNewLauncherIcon(
        'ic_launcher',
        androidManifestFile,
      );
      expect(androidManifestFile.readAsStringSync(), equals(expectedManifest));
    });
  });

  test(
      'Transforming manifest with special newline characters should leave special newline characters untouched',
      () async {
    final String inputManifest =
        getAndroidManifestExample('android:icon="@mipmap/ic_launcher"')
            .replaceAll('\n', '\r\n');
    final String expectedManifest = inputManifest;

    await withTempFile('AndroidManifest.xml', (File androidManifestFile) async {
      androidManifestFile.writeAsStringSync(inputManifest);
      await android.overwriteAndroidManifestWithNewLauncherIcon(
        'ic_launcher',
        androidManifestFile,
      );
      expect(androidManifestFile.readAsStringSync(), equals(expectedManifest));
    });
  });
}

Future<void> withTempFile(String fileName, Function block) async {
  final Directory tempDir = Directory.systemTemp.createTempSync();
  final File file = File('${tempDir.path}/$fileName')..createSync();
  if (!file.existsSync()) {
    fail('Could not create temp test file ${file.path}');
  }
  try {
    await block(file);
  } finally {
    file.deleteSync();
  }
}

String getAndroidManifestExample(String iconLine) {
  return '''
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
    package="com.example.myapplication">

    <application
        android:allowBackup="true"
        $iconLine
        android:label="@string/app_name"
        android:roundIcon="@mipmap/ic_launcher_round"
        android:supportsRtl="true"
        android:theme="@style/AppTheme">
        <activity
            android:name=".MainActivity"
            android:label="@string/app_name"
            android:theme="@style/AppTheme.NoActionBar">
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />

                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
        </activity>
    </application>

</manifest>
  '''
      .trim();
}


lib/constants.dart
// ignore_for_file: public_member_api_docs

import 'package:path/path.dart' as path;

/// Relative path to android resource folder
String androidResFolder(String? flavor) =>
    "android/app/src/${flavor ?? 'main'}/res/";

/// Relative path to android colors.xml file
String androidColorsFile(String? flavor) =>
    "android/app/src/${flavor ?? 'main'}/res/values/colors.xml";

const String androidManifestFile = 'android/app/src/main/AndroidManifest.xml';
const String androidGradleFile = 'android/app/build.gradle';
const String androidLocalPropertiesFile = 'android/local.properties';

/// Relative path to flutter.gradle from flutter sdk path
const String androidFlutterGardlePath =
    'packages/flutter_tools/gradle/flutter.gradle';

/// Default min_sdk value for android
/// https://github.com/flutter/flutter/blob/master/packages/flutter_tools/gradle/flutter.gradle#L35-L37
const int androidDefaultAndroidMinSDK = 21;
const String androidFileName = 'ic_launcher.png';
const String androidAdaptiveForegroundFileName = 'ic_launcher_foreground.png';
const String androidAdaptiveBackgroundFileName = 'ic_launcher_background.png';
String androidAdaptiveXmlFolder(String? flavor) =>
    androidResFolder(flavor) + 'mipmap-anydpi-v26/';
const String androidDefaultIconName = 'ic_launcher';

const String iosDefaultIconFolder =
    'ios/Runner/Assets.xcassets/AppIcon.appiconset/';
const String iosAssetFolder = 'ios/Runner/Assets.xcassets/';
const String iosConfigFile = 'ios/Runner.xcodeproj/project.pbxproj';
const String iosDefaultIconName = 'Icon-App';

// web
/// favicon.ico size
const int kFaviconSize = 16;

/// Relative web direcotry path
String webDirPath = path.join('web');

/// Relative web icons directory path
String webIconsDirPath = path.join(webDirPath, 'icons');

/// Relative web manifest.json file path
String webManifestFilePath = path.join(webDirPath, 'manifest.json');
// TODO(RatakondalaArun): support for other images formats
/// Relative favicon.png path
String webFaviconFilePath = path.join(webDirPath, 'favicon.png');

/// Relative index.html file path
String webIndexFilePath = path.join(webDirPath, 'index.html');

/// Relative pubspec.yaml path
String pubspecFilePath = path.join('pubspec.yaml');

// Windows
/// Relative path to windows directory
String windowsDirPath = path.join('windows');

/// Relative path to windows resources directory
String windowsResourcesDirPath =
    path.join(windowsDirPath, 'runner', 'resources');

/// Relative path to windows icon file path
String windowsIconFilePath = path.join(windowsResourcesDirPath, 'app_icon.ico');

/// Default windows icon size for flutter
///
const int windowsDefaultIconSize = 48;

// MacOS

/// Relative path to macos folder
final macOSDirPath = path.join('macos');

/// Relative path to macos icons folder
final macOSIconsDirPath =
    path.join(macOSDirPath, 'Runner', 'Assets.xcassets', 'AppIcon.appiconset');

/// Relative path to macos contents.json
final macOSContentsFilePath = path.join(macOSIconsDirPath, 'Contents.json');

const String errorMissingImagePath =
    'Missing "image_path" or "image_path_android" + "image_path_ios"'
    ' + "image_path_web" within configuration';
const String errorMissingPlatform =
    'No platform specified within config to generate icons for.';
const String errorMissingRegularAndroid =
    'Adaptive icon config found but no regular Android config. '
    'Below API 26 the regular Android config is required';
const String errorMissingMinSdk =
    'Cannot not find minSdk from android/app/build.gradle or android/local.properties'
    ' Specify minSdk in your flutter_launcher_config.yaml with "min_sdk_android"';
const String errorIncorrectIconName =
    'The icon name must contain only lowercase a-z, 0-9, or underscore: '
    'E.g. "ic_my_new_icon"';

const String errorWebCustomLocationNotSupported =
    'Icon generation for web does not support specification of a custom '
    'icon location! Please manually save a copy of the prefered icon theme '
    '(perhaps by making a copy and renaming it). Custom locations are not '
    'supported.';

String warningPlatformDirectoryMissing(String platformName) =>
    generateWarning(' Not generating icons for $platformName.'
        ' Platform $platformName is in config, but there is no'
        ' matching directory. If you want to generate icons for'
        ' $platformName, make sure you have support for $platformName'
        ' enabled and have run `flutter create .`'
        ' in the root directory of your project.');

const String currentVersion = '0.0.0';
const String introMessage = '''
  ════════════════════════════════════════════
     FLUTTER LAUNCHER ICONS (v$currentVersion)                               
  ════════════════════════════════════════════
  ''';

const String fileOption = 'file';
const String helpFlag = 'help';
const String defaultConfigFile = 'flutter_launcher_icons.yaml';
const String flavorConfigFilePattern = r'^flutter_launcher_icons-(.*).yaml$';

